### PR TITLE
[connection.py] bug fix in funciton 'gethost_addrinfo'

### PR DESCRIPTION
**Pull request by XiaoliChan**
This is a bug from `connection.py`

For more details: https://github.com/mpgn/CrackMapExec/issues/128

### CODE REVIEW

**Comment by XiaoliChan** [comment]
Looks like it will break `ldap`, still need to check

**Comment by XiaoliChan** [comment]
Reference
https://github.com/Porchetta-Industries/CrackMapExec/commit/27552fab02a3962ef53e9f9a404017037aaabcb2

**Comment by XiaoliChan** [comment]
hmm, something wrong with ipv6 in local link

**Comment by XiaoliChan** [comment]
StringBinding does not support local link ipv6, if it failed, return None
image

File: cme/connection.py [hunk 1]
```diff
@@ -58,6 +58,8 @@ def dcom_FirewallChecker(iInterface, timeout):
                 break
             elif iInterface.is_fqdn() and binding.upper().find(iInterface.get_target().upper().partition('.')[0]) >= 0:
                 stringBinding = 'ncacn_ip_tcp:%s%s' % (iInterface.get_target(), bindingPort)
+    if "stringBinding" not in locals():
+        return False, None
```
Comment (mpgn) [inline cme/connection.py:62]: return True, None

If stringbinding is failing because of ipv6, it's not relevant to show an error message not related to firewall

File: cme/connection.py [hunk 2]
```diff
@@ -26,12 +27,12 @@ def gethost_addrinfo(hostname):
     try:
         for res in getaddrinfo( hostname, None, AF_INET6, SOCK_DGRAM, IPPROTO_IP, AI_CANONNAME):
             af, socktype, proto, canonname, sa = res
+        host = canonname if ip_address(sa[0]).is_link_local else sa[0]
```
Comment (mpgn) [inline cme/connection.py:30]: host is defined here regarding sa

File: cme/connection.py [hunk 3]
```diff
@@ -26,12 +27,12 @@ def gethost_addrinfo(hostname):
...
     except socket.gaierror:
         for res in getaddrinfo( hostname, None, AF_INET, SOCK_DGRAM, IPPROTO_IP, AI_CANONNAME):
             af, socktype, proto, canonname, sa = res
-    if canonname == "":
-        return sa[0]
-    return canonname
+        host = canonname if not sa[0] else sa[0]
```
Comment (mpgn) [inline cme/connection.py:34]: but host is defined again regarding sa, which condition is the right one

File: cme/connection.py (hunk 2 above)
Comment (XiaoliChan) [reply cme/connection.py:30]: OK, Let me explain this, so I have check the history commit, if the IPV6 is local link address like `fe80::a6d0:523f:732d:9c46%eth0`, then it must return `canonname`, in `sa`, the "%eth0" will be removed, and make CME didn't work.

But for real ipv6 address, like "240c::6666", because it is not a local link address, we don't need to care about `%eth0`, so just return `sa`

File: cme/connection.py (hunk 3 above)
Comment (XiaoliChan) [reply cme/connection.py:34]: In this case, if it can't resolve the hostname, the return the `canonname`, otherwise, this should always return `sa`

File: cme/connection.py (hunk 1 above)
Comment (XiaoliChan) [reply cme/connection.py:62]: You are right, should tell the users about the ipv6 error.

But, my idea is, if it `False`, which means it can't go into the firewall check in the next step https://github.com/mpgn/CrackMapExec/pull/129/files/53d7d06fcf51acb0fd0264f5e761d35904b4ab97..a13e45bb56d0c1e40561e5526389ad2362bc18d9#diff-cc496d0ee13e4bf89b3b442c4a4b996889f50c2c302c94de40df4e3ef34e2b6cR64

So, after this PR, I just need to add an condition `if stringBinding` in https://github.com/mpgn/CrackMapExec/blob/master/cme/protocols/smb/wmiexec.py#L80C43-L80C43

File: cme/connection.py (hunk 2 above)
Comment (mpgn) [reply cme/connection.py:30]: i understand but 5 line after you redefine host variable so i don't understand the reason of this line

File: cme/connection.py (hunk 1 above)
Comment (mpgn) [reply cme/connection.py:62]: why not on this PR ? :)

File: cme/connection.py (hunk 2 above)
Comment (XiaoliChan) [reply cme/connection.py:30]: > i understand but 5 line after you redefine host variable so i don't understand the reason of this line

`host = canonname if not sa[0] else sa[0]` this is for Ipv4 (`AF_INET`)

Comment (mpgn) [reply cme/connection.py:30]: damm, i missread the indentation, sorry :D

File: cme/connection.py (hunk 1 above)
Comment (XiaoliChan) [reply cme/connection.py]: Done!

File: cme/modules/rdp.py [hunk 4]
```diff
@@ -248,11 +248,15 @@ def __init__(self, context, connection, timeout):
         iInterface = self.__dcom.CoCreateInstanceEx(wmi.CLSID_WbemLevel1Login, wmi.IID_IWbemLevel1Login)
         if self.__currentprotocol == "smb":
             flag, self.__stringBinding =  dcom_FirewallChecker(iInterface, self.__timeout)
-            if not flag:
-                self.logger.fail(f'WMIEXEC: Dcom initialization failed on connection with stringbinding: "{self.__stringBinding}", please increase the timeout with the module option "DCOM-TIMEOUT=10". If it\'s still failing maybe something is blocking the RPC connection, try "METHOD=smb"')
+            if not flag or not self.__stringBinding:
```
Comment (mpgn) [inline cme/modules/rdp.py:251]: the logic should be "show fail if this is related to firewall but if this is related to ipv6 show error as debug"

Comment (XiaoliChan) [reply cme/modules/rdp.py:251]: I see

Comment (XiaoliChan) [reply cme/modules/rdp.py:251]: OH, I am changing `rdp.py` this will make conflict with PR #126

File: cme/modules/rdp.py
Comment (XiaoliChan) [inline cme/modules/rdp.py:1]: Need to merge #126 first, my fault.

**Comment by XiaoliChan** [comment]
Done, I just merged PR #126 into this PR and did some self review.

File: cme/modules/rdp.py [hunk 5]
```diff
@@ -232,28 +233,39 @@ def __init__(self, context, connection, timeout):
...
+                self.__aesKey,
+                oxidResolver=True,
+                doKerberos=self.__doKerberos,
+                kdcHost=self.__kdcHost,
+            )
+
+            iInterface = self.__dcom.CoCreateInstanceEx(wmi.CLSID_WbemLevel1Login, wmi.IID_IWbemLevel1Login)
+            if self.__currentprotocol == "smb":
+                flag, self.__stringBinding =  dcom_FirewallChecker(iInterface, self.__timeout)
+                if not flag or not self.__stringBinding:
```
Comment (mpgn) [inline cme/modules/rdp.py:253]: `if no flag` is enough

File: cme/protocols/smb.py [hunk 6]
```diff
@@ -1233,19 +1233,24 @@ def wmi(self, wmi_query=None, namespace=None):
             )
             iInterface = dcom.CoCreateInstanceEx(CLSID_WbemLevel1Login,IID_IWbemLevel1Login)
             flag, stringBinding =  dcom_FirewallChecker(iInterface, self.args.dcom_timeout)
-            if not flag:
-                self.logger.fail(f'WMI Query: Dcom initialization failed on connection with stringbinding: "{stringBinding}", please increase the timeout with the option "--dcom-timeout". If it\'s still failing maybe something is blocking the RPC connection')
+            if not flag or not stringBinding:
```
Comment (mpgn) [inline cme/protocols/smb.py:1236]: `if no flag` is enough

File: cme/protocols/smb/wmiexec.py [hunk 7]
```diff
@@ -75,10 +75,18 @@ def __init__(
         )
         iInterface = self.__dcom.CoCreateInstanceEx(wmi.CLSID_WbemLevel1Login, wmi.IID_IWbemLevel1Login)
         flag, self.__stringBinding =  dcom_FirewallChecker(iInterface, self.__timeout)
-        if flag is False:
-            self.logger.fail(f'WMIEXEC: Dcom initialization failed on connection with stringbinding: "{self.__stringBinding}", please increase the timeout with the option "--dcom-timeout". If it\'s still failing maybe something is blocking the RPC connection, try another exec method')
+        if not flag or not self.__stringBinding:
```
Comment (mpgn) [inline cme/protocols/smb/wmiexec.py:78]: `if no flag` is enough

File: cme/modules/rdp.py (hunk 5 above)
Comment (XiaoliChan) [reply cme/modules/rdp.py:253]: well, that mind be wrong, because it return True when cause by ipv6

Comment (XiaoliChan) [reply cme/modules/rdp.py:253]: https://github.com/mpgn/CrackMapExec/pull/129/commits/0ef36b5c385138c4cb41efda773f697fe1557136

Comment (mpgn) [reply cme/modules/rdp.py:253]: arf it's the indentation again that is playing with me, i thought if not flag or not self.__stringBinding: and if not self.__stringBinding: were on the same leve.

**Review by mpgn — APPROVED** [review]

File: cme/modules/rdp.py (hunk 5 above)
Comment (XiaoliChan) [reply cme/modules/rdp.py:253]: So, it should be right.

File: cme/modules/rdp.py [hunk 8]
```diff
@@ -85,11 +85,10 @@ def on_admin_login(self, context, connection):
             context.log.info("Executing over WMI(ncacn_ip_tcp)")
             try:
                 wmi_rdp = rdp_WMI(context, connection, self.dcom_timeout)
-            except Exception as e:
-                context.log.fail(f"Unexpected wmi error: {str(e)}")
-                wmi_rdp._rdp_WMI__dcom.disconnect()
+            except:
+                pass
```
Comment (NeffIsBack) [inline cme/modules/rdp.py:89]: Why did you remove the logging and disconnect?

File: cme/protocols/smb.py [hunk 9]
```diff
@@ -1233,19 +1233,24 @@ def wmi(self, wmi_query=None, namespace=None):
...
+                    error_msg = "WMI Query: Dcom initialization failed: can't get target stringbinding, maybe cause by IPv6 or any other issues, please check your target again"
+                
+                self.logger.fail(error_msg) if not flag else self.logger.debug(error_msg)
                 # Make it force break function
-                dcom.disconnect()
-                return False
+                try:
+                    dcom.disconnect()
+                except:
+                    pass
```
Comment (NeffIsBack) [inline cme/protocols/smb.py:1247]: Same here

File: cme/protocols/smb/mmcexec.py [hunk 10]
```diff
@@ -100,38 +100,44 @@ def __init__(self, host, share_name, username, password, domain, smbconnection,
...
+            if not self.__stringBinding:
+                error_msg = "MMCEXEC: Dcom initialization failed: can't get target stringbinding, maybe cause by IPv6 or any other issues, please check your target again"
+            
+            self.logger.fail(error_msg) if not flag else self.logger.debug(error_msg)
+            try:
                 # Make it force break function
                 self.__dcom.disconnect()
-            iMMC = IDispatch(iInterface)
+            except:
+                pass
```
Comment (NeffIsBack) [inline cme/protocols/smb/mmcexec.py:118]: Same here

File: cme/protocols/smb/mmcexec.py [hunk 11]
```diff
@@ -168,13 +174,16 @@ def execute(self, command, output=False):
...
+            dispParams = DISPPARAMS(None, False)
+            dispParams["rgvarg"] = NULL
+            dispParams["rgdispidNamedArgs"] = NULL
+            dispParams["cArgs"] = 0
+            dispParams["cNamedArgs"] = 0
 
-        self.__quit[0].Invoke(self.__quit[1], 0x409, DISPATCH_METHOD, dispParams, 0, [], [])
+            self.__quit[0].Invoke(self.__quit[1], 0x409, DISPATCH_METHOD, dispParams, 0, [], [])
+        except:
+            pass
```
Comment (NeffIsBack) [inline cme/protocols/smb/mmcexec.py:186]: Same here. If you expect errors to happen imo we should handle them properly. Or is there a good reason to not properly catch the exception?

File: cme/protocols/smb/wmiexec.py [hunk 12]
```diff
@@ -75,10 +75,18 @@ def __init__(
...
+            if not self.__stringBinding:
+                error_msg = "WMIEXEC: Dcom initialization failed: can't get target stringbinding, maybe cause by IPv6 or any other issues, please check your target again"
+            
+            self.logger.fail(error_msg) if not flag else self.logger.debug(error_msg)
             # Make it force break function
-            self.__dcom.disconnect()
+            try:
+                self.__dcom.disconnect()
+            except:
+                pass
```
Comment (NeffIsBack) [inline cme/protocols/smb/wmiexec.py:89]: Same as above

File: cme/modules/rdp.py (hunk 8 above)
Comment (XiaoliChan) [reply cme/modules/rdp.py:89]: This can explain, already in line 264
https://github.com/mpgn/CrackMapExec/blob/e3c734c186bfcd7e9935cd63556bc6c060bf397c/cme/modules/rdp.py#L264

File: cme/modules/rdp.py [hunk 13]
```diff
@@ -110,7 +108,10 @@ def on_admin_login(self, context, connection):
                         else:
                             context.log.fail(str(e))
                         pass
-                    wmi_rdp._rdp_WMI__dcom.disconnect()
+            try:
+                wmi_rdp._rdp_WMI__dcom.disconnect()
+            except:
+                pass
```
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py:114]: Sorry, `try except` in here is unnecessary

File: cme/modules/rdp.py [hunk 14]
```diff
@@ -232,28 +233,39 @@ def __init__(self, context, connection, timeout):
...
+                    self.__dcom.disconnect()
+            self.__iWbemLevel1Login = wmi.IWbemLevel1Login(iInterface)
+        except Exception as e:
+            self.logger.fail(f'Unexpected wmi error: {str(e)}, please try to use "-o" with "METHOD=smb"')
+            try:
                 self.__dcom.disconnect()
-                return
-        self.__iWbemLevel1Login = wmi.IWbemLevel1Login(iInterface)
+            except:
+                pass
```
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py:268]: Sorry, `try except` in here is unnecessary

File: cme/protocols/smb.py (hunk 9 above)
Comment (XiaoliChan) [reply cme/protocols/smb.py:1247]: Sorry, you are right, `try except` in here is unnecessary

File: cme/protocols/smb/mmcexec.py (hunk 10 above)
Comment (XiaoliChan) [reply cme/protocols/smb/mmcexec.py:118]: Sorry, you are right, try except in here is unnecessary

File: cme/protocols/smb/mmcexec.py (hunk 11 above)
Comment (XiaoliChan) [reply cme/protocols/smb/mmcexec.py:186]: Yes, you are right, shouldn't pass it.

File: cme/protocols/smb/wmiexec.py (hunk 12 above)
Comment (XiaoliChan) [reply cme/protocols/smb/wmiexec.py:89]: Sorry, you are right, `try except` in here is unnecessary

**Comment by XiaoliChan** [comment]
@NeffIsBack So sorry about that ugly `try except`, because some `try except` I just playing with the ipv6 error and I forgot remove

File: cme/protocols/smb/mmcexec.py (hunk 11 above)
Comment (XiaoliChan) [reply cme/protocols/smb/mmcexec.py]: Fixed

File: cme/protocols/smb/mmcexec.py (hunk 10 above)
Comment (XiaoliChan) [reply cme/protocols/smb/mmcexec.py]: Fixed

File: cme/protocols/smb.py (hunk 9 above)
Comment (XiaoliChan) [reply cme/protocols/smb.py]: Fixed

File: cme/modules/rdp.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py]: Fixed

File: cme/modules/rdp.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py]: Fixed

File: cme/modules/rdp.py (hunk 8 above)
Comment (NeffIsBack) [reply cme/modules/rdp.py:89]: I think the try except is unnecessary then right? It should get handled in the method itself

Comment (XiaoliChan) [reply cme/modules/rdp.py]: Sure!

Comment (XiaoliChan) [reply cme/modules/rdp.py]: Fixed!

**Comment by XiaoliChan** [comment]
Confirm all the reviews have been solved.

**Review by mpgn — APPROVED** [review]
